errors.go

```diff
@@ -0,0 +1,244 @@
+package flux
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// Some error code constant, ideally we want define common platform codes here
+// projects on use platform's error, should have their own central place like this.
+// Any time this set of constants changes, you must also update the swagger for Error.properties.code.enum.
+const (
+	EInternal            = "internal error"
+	ENotFound            = "not found"
+	EConflict            = "conflict"             // action cannot be performed
+	EInvalid             = "invalid"              // validation failed
+	EUnprocessableEntity = "unprocessable entity" // data type is correct, but out of range
+	EEmptyValue          = "empty value"
+	EUnavailable         = "unavailable"
+	EForbidden           = "forbidden"
+	ETooManyRequests     = "too many requests"
+	EUnauthorized        = "unauthorized"
+	EMethodNotAllowed    = "method not allowed"
```

Comment:
These error code seem very specific to HTTP and that does not apply well to Flux as the repo itself does not interact with HTTP APIs.

Reply:
I agree on some of them. `EMethodNotAllowed` is a bit _too_ http specific. Forbidden and unauthorized are decently generic even if they map directly to HTTP. Too many requests is technically generic even if it is directly mapped to http.

One possible issue though is, a little by necessity, these need to be the same as influxdb. I guess we can remove the ones that aren't used or are more egregious like method not allowed, but I think that's a judgment call.

Reply:
Why do we need to the same codes as influxdb? That is not clear to me.

Also it looks like the methods below have been implemented by using type assertions on anonymous interfaces so copying the types should also be safe. Not translation should be needed from errors originating in Flux vs errors originating in Influxdb.

Reply:
It's because of serialization. If we don't have the same codes as influxdb, then they won't be able to read our error codes properly.

They'll do something like:

```go
if influxdb.ErrorCode(err) == influxdb.EForbidden {
    // ...
}
```

And even if we returned `EForbidden`, it won't recognize it. Additionally, JSON serialization, which is very possible that this will go through, will cause a `flux.Error` to get reconstructed as an `influxdb.Error` so it wouldn't be possible to differentiate the two by what type they are. We cannot make the assumption a `flux.Error` will stay a `flux.Error`.

Reply:
I see, I had assumed that these error were meant to be opaque to the code, but looking through the influxdb code I see that is not the case. I think we should add a comment explicitly stating that since common Go error handling is that errors are opaque.

Is there a different approach we can take that doesn't tightly couple error handling in Flux with how InfluxDB does error handling?


As an aside:

> We cannot make the assumption a flux.Error will stay a flux.Error.

This is a good thing in my mind. It means the errors are defined by their behavior and not their representation.

Reply:
I think it would be best to take the core of the errors library from influxdb and put it into its own separate and versioned package. The utility methods that influxdb has don't have to come along with it for simplicity and to ensure the interface doesn't change. That way we have no dependency on influxdb and we can type alias all of the functions we want or any new ones.

The error codes are supposed to be generic similar to HTTP (although the codes are not supposed to be HTTP) so they aren't supposed to be influxdb specific. There's a note about adding them to the swagger doc, but that could be changed to a go generate command that reads a file from a dependency in influxdb.

Reply:
For clarity, I don't think using the same error handling would tightly couple the two. Once it is its own library, it stands on its own and it just happens to be that the two use the same method of reporting errors.

Reply:
I see two ways this could work:

1. Couple the error handling in which case it makes sense to create a separate package that flux and influxdb both depend on. Otherwise the divergence would create a nightmare of bugs.
2. Decouple the error handling, meaning do not use the same error codes or types, but just take the same basic strategy. Flux can export its error codes and InfluxDB can make an explicit translation between them if needed. I would rather Flux errors be opaque to InfluxDB with a few explicit exceptions. This way packages consuming Flux (i.e. InfluxDB) can make decisions about what Flux error codes mean and if they will take action on them, otherwise they can treat them as opaque. The only current use case will be for Flux to report errors as either system errors or user errors, which InfluxDB will inspect to update appropriate statistics.

I prefer option 2 as it reduces the surface area of the Flux Go API. 

I have a few concerns over how option 1 could play out in confusing and negative ways. 

* Since InfluxDB and Flux will use the same error codes, the set of error codes will be the union of the two needs. In addition to the error code being part of the  InfluxDB public API this means that changes need to be strictly controlled and may not be possible to change in some cases. These two combined leads to a tight coupling that makes it hard to make changes to either project.
* All consumers of Flux as a library will need to make a decision about how to handle the error codes Flux can produce. As such that set of error codes should be as small as possible, including influxdb specific codes is confusing and error prone to consumers of Flux.
* This tight coupling can cause very hard to debug and unexpected issues. For example given that InfluxDB has logic that conditions on the error codes of various errors. And given that Flux produces errors with error codes that InfluxDB consumes. A change to errors either in Flux code or anywhere in the call stack between Flux and InfluxDB can trigger different logic in InfluxDB. These kinds of changes will have very indirect and hard to predict behaviors as the logic being affected is far away from the changes being made. 

Said another way, in general the practice of inspecting errors creates the possibility that changes can effect code that is loosely related (i.e. way up the call stack). As such I want to tightly control that surface area so it is more clear how said changes effect the code base as a whole.

Reply:
Ok. I think this means the interfaces aren't really needed then because we would be saying the API is completely different so there's no expectation the values are the same. I think that means we'll need to add a method to influxdb so if it sees a `*flux.Error` it will translate it to an `*influxdb.Error`. I took another look at the source blog post for our system ([here](https://middlemost.com/failure-is-your-domain/) and that's what it said to do so I think that's a good solution.

Reply:
> Since InfluxDB and Flux will use the same error codes, the set of error codes will be the union of the two needs. In addition to the error code being part of the InfluxDB public API this means that changes need to be strictly controlled and may not be possible to change in some cases. These two combined leads to a tight coupling that makes it hard to make changes to either project.

I don't think this will be a problem. Error codes are for groups of problems that inform consumers of the library what to do. Except the `method not available` I think each of the error codes in errors.go today are good groupings and each are applicable for flux. 

I think method not available should be removed and I'll make a PR to do so.

What groupings of errors do you imagine are not covered?  Flux is integral to influxdb and if influxdb does not support an error group flux needs, then, it absolutely should be added.

>    All consumers of Flux as a library will need to make a decision about how to handle the error codes Flux can produce. As such that set of error codes should be as small as possible, including influxdb specific codes is confusing and error prone to consumers of Flux.

I don't really see any error groupings that would not be appropriate for flux except method not allowed.

>    This tight coupling can cause very hard to debug and unexpected issues. For example given that InfluxDB has logic that conditions on the error codes of various errors. And given that Flux produces errors with error codes that InfluxDB consumes. A change to errors either in Flux code or anywhere in the call stack between Flux and InfluxDB can trigger different logic in InfluxDB. These kinds of changes will have very indirect and hard to predict behaviors as the logic being affected is far away from the changes being made.

Ultimately, the behavior that would change in influxdb from error codes would be the HTTP status code from /api/v2/query.  I think it is quite reasonable for flux to govern /api/v2/query